L17

CaptainPhilipp opened the pull request.

CaptainPhilipp committed:
sort stations in route

CaptainPhilipp committed:
прост))

CaptainPhilipp committed:
nested carriages for train

CaptainPhilipp committed:
arrival and departure time

CaptainPhilipp committed:
stylefix

CaptainPhilipp committed:
prepare to changes

CaptainPhilipp committed:
partial + time

CaptainPhilipp committed:
prepare model

CaptainPhilipp committed:
Tickets, Route search

CaptainPhilipp committed:
forms

CaptainPhilipp committed:
Service

CaptainPhilipp committed:
new ticket, and some other restructs

CaptainPhilipp committed:
fixes

vkurennov commented on app/controllers/tickets_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+  private
+
+  def set_ticket_context(params_)
```
Особого смысла передавать параметры как аргумент нет, ведь это же обычный инстанс-метод, он и так имеет доступ к методу `params`.

CaptainPhilipp replied:
я сделал так, без фильтра и с передачей `params` через аргументы, тк в одном экшене эти данные берутся через `params.require(:ticket)`, в другом - просто из `params` (тк параметры передаются туда уже не из формы, а через роут)

собсно потому аргументу и добавил в конце `_` что бы указать его возможное отличие от `params`

vkurennov replied:
ок, принято

vkurennov commented on app/controllers/tickets_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def create
+    @ticket = Ticket.new(ticket_params)
+    set_ticket_context(ticket_params)
```
Вот это нужно будет только для `new`, значит и вызывать это можно в `else`

vkurennov commented on app/controllers/trains_controller.rb:
```diff
@@ -51,12 +52,22 @@ def destroy
     redirect_to trains_url, notice: 'Train was successfully destroyed.'
   end
 
+  def search
```
Поиск лучше отдельным контроллером сделать, имхо. Да, формально мы ищем поезда, но сам по себе поиск - это отдельная страница и самостоятельный ресурс. Он может усложняться, поэтому лучше его отделить от контроллера поездов.

vkurennov commented on app/helpers/trains_helper.rb:
```diff
@@ -1,2 +1,6 @@
 module TrainsHelper
```
Хелперы нужны для видов, именно для этого и есть папка helpers, если нужен какой-то модуль только для контроллера, то лучше его поместить в папку `controllers/concerns`

CaptainPhilipp replied:
эту штуку я перенёс сюда из контроллера т.к. она использовалась не только в самом контроллере, но и в его вьюхах.

vkurennov replied:
А зачем во вьюхах стронг-параметры?

CaptainPhilipp replied:
а, точно, я же их уже убрал из вьюх...
забыл перенеси метод обратно в контроллер под приват. ему там уютнее.

vkurennov commented on app/models/railway_station.rb:
```diff
@@ -5,9 +5,31 @@ class RailwayStation < ApplicationRecord
-    rel_railway_stations_routes.
-      find_by(route_id: route.id).
-      update(sort_key: sort_key)
+  scope :ordered_by_route, lambda {
```
Сейчас принято указывать `->`, а не `lambda`

CaptainPhilipp replied:
а меня вроде рубокоп когда-то поправлял, говорил для многострочников юзать lambda. может что-то не уловил... посмотрю.

vkurennov replied:
Хм, может быть, но тогда и do end надо использовать, а не фигруные скобки

CaptainPhilipp replied:
проверил - `lambda do \n end`  кидает ошибку мол пытаюсь сделать прок без блока.
а на `-> do \n end` коп жалуется. на `-> { \n }` - стайлгайд не доволен.

vkurennov replied:
Ок, пусть так остается

CaptainPhilipp commented on config/routes.rb:
```diff
@@ -1,5 +1,17 @@
+  end
+
+  resources :tickets, only: %i(create show index) do
+    get 'new/:train_id/:departure_id/:arrival_id', to: 'tickets#new', on: :collection, as: :new
```
выбирая между передачей параметров через создание формы с `hidden_field` для каждой строки каждого поезда (т.е. в html на каждый найденный поезд будут дублироваться одни и те же `hidden_field_tag` с `departure_id`, и `arrival_id`)
и кастомным роутом, содержащим все необходимые аргументы, я выбрал роут.
Может есть вариант получше в данной ситуации?

vkurennov replied:
С роутами лучше, имхо

CaptainPhilipp commented on app/models/railway_station.rb:
```diff
@@ -5,9 +5,31 @@ class RailwayStation < ApplicationRecord
-    rel_railway_stations_routes.
-      find_by(route_id: route.id).
-      update(sort_key: sort_key)
+  scope :ordered_by_route, lambda {
```
а с do end вообще мистика:
Еще в BlackJack писал `lambda do /n end`, и всё правильно, 'на стиле' - интерпретатор доволен, коп доволен, стайлгайд одобряет.
А уже недавно в рельсах на скоупах вдруг полетели ошибки - требует скобок даже на многострочные блоки.
вообще, это на столько странно что сам себе сейчас не верю, буду перепроверять.
Но точно помню что он грубо ругался > я удивился, поменял > работает.

CaptainPhilipp committed:
after-review little fixes

CaptainPhilipp committed:
Searches controller

CaptainPhilipp committed:
Oops =P

методом научного тыка пытался понять в каких случаях рубокоп не одобряет префикс _set, но забыл поправить испорченное >_<

CaptainPhilipp committed:
root page, StaticPages controller

CaptainPhilipp committed:
train carriages full shallow

CaptainPhilipp committed:
Carriages Fix

CaptainPhilipp committed:
show ticket

CaptainPhilipp committed:
home_path

vkurennov commented on app/controllers/searches_controller.rb:
```diff
@@ -0,0 +1,16 @@
+class SearchesController < ApplicationController
+  def trains_by_terminal_stations
```
Не стоили делать кастомный экшн, да еще и с таким длинным и неудобным названием. Досаточно было, например, new или show.

vkurennov commented on config/routes.rb:
```diff
@@ -1,5 +1,21 @@
+    get 'new/:train_id/:departure_id/:arrival_id', to: 'tickets#new', on: :collection, as: :new
+  end
+
+  namespace :searches do
```
Это не неймспейс должен быть, а ресурс. И раз он в одном экземпляре существует, то лучше его и объявлять как `resource :search`

CaptainPhilipp committed:
rename, restruct routes

CaptainPhilipp committed:
Revert "rename, restruct routes"

This reverts commit 92065c504f5185ebce4e717151e7c2cda1ca5372.
